### PR TITLE
fix(trtllm): return 4xx for client-fault embedding load failures

### LINKED ISSUE #14749: Multimodal processing failures return a sanitized 500 even when the client's URL is at fault
## Problem

`components/src/dynamo/trtllm/request_handlers/handler_base.py` raises `RuntimeError` for a multimodal request it cannot process:

```python
raise RuntimeError(
    "Failed to process multimodal request. Check server logs for details. "
    "Common issues: missing allowed_local_media_path configuration, "
    "file not found, or file outside allowed directory."
)
```

`RuntimeError` maps to `Backend(Unknown)` in `py_err_to_dynamo` (`lib/bindings/python/rust/backend.rs`), so the frontend answers `500` and replaces the body with a sanitized "Internal server error". The client loses the only actionable text there is.

"file not found" and "file outside allowed directory" are verdicts on the **client's** URL, so at least those cases warrant a 4xx.

## Complication

The message conflates three causes with different owners:

- `allowed_local_media_path` not configured — server misconfiguration, arguably 5xx
- file not found — client's bad URL, 4xx
- file outside allowed directory — client's bad URL, 4xx

A blanket retype to `InvalidArgument` would label the misconfiguration case a client error. The cases probably want splitting before the type is changed.

## Context

The sibling guard ~20 lines above was retyped to `InvalidArgument` in #14467; this one was deliberately left alone because of the ambiguity above.

### PR DESCRIPTION
**Pull request by Bossixd**
## Summary:

A multimodal request on trtllm whose pre-computed embedding file (`.safetensors`) could not be loaded got a sanitized HTTP `500`, even when the client's URL was the problem. In `components/src/dynamo/trtllm/multimodal_processor.py`, `load_tensor_from_path_or_url()` raised `RuntimeError`, `process_openai_request()` swallowed it into `return None`, and `handler_base.py`  raised a generic `RuntimeError`. That maps to `Backend(Unknown)`, so the frontend gets a `500 "Internal server error"` and the reason is not propagated.

This PR classifies each failure by who caused it. Client-fault cases raise `HttpStatusError(400, ...)` and reach the frontend as a `400` with an actionable message. The server-configuration case stays a `5xx`.

## Details:

**`load_tensor_from_path_or_url`** (`components/src/dynamo/trtllm/multimodal_processor.py`)
- `RuntimeError` → `HttpStatusError(400)` for the client-related error cases:
  - embedding file not found
  - file outside `allowed_local_media_path`
  - file over the size limit
  - unsupported URL scheme
  - not `.safetensors` / unsafe tensor format
- Client-facing messages only shows the client's own input, never the server's path or the allowed directory. The containment check still runs before the existence check, so sending a "not found" file cannot be used to list files outside the allowed directory.
- `allowed_local_media_path` unset → `HttpStatusError(500)`. The issue [#14749](https://github.com/ai-dynamo/dynamo/issues/14749) leans towards this case being ambiguous, but I kept it `5xx` as proposed in the issue because it is the server's configuration causing it, not the client's request.

**`process_openai_request`**
- The embedding block now re-raises `HttpStatusError` before its generic `except Exception`, mirroring the existing image logic. Previously every embedding error just did `return None`.
- Dict embeddings missing the `mm_embeddings` key → `400`. The echoed path is bounded via `describe_media_source`.

**Scope notes**
- `load_tensor_from_path_or_url` is also called by the EPD encode worker (`encode_helper.py`), so its errors now carry a status there too. The encode→prefill hop is not verified end-to-end.
- As [#14467](https://github.com/ai-dynamo/dynamo/pull/14467) notes, `stream: true` clients still see `200` followed by an SSE error frame unless `DYN_HTTP_PRE_COMMIT_ERROR_PEEK_MS` is set. The `4xx` reaches non-streaming clients.

## Validation:

- Ran `components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py` in `nvcr.io/nvidia/ai-dynamo/tensorrtllm-runtime:1.4.2`, with this branch mounted over the image via `PYTHONPATH=/src/components/src`, on an RTX 3050 Ti: **26 tests passed**.
- Negative control: with `multimodal_processor.py` reverted to `main` branch, **6 of the new tests fail**. These are the five `400` cases and the unconfigured `500` case, which confirms the tests catch the original behaviour. `test_embedding_valid_file_loads` pass either way, as expected, since they test existing behaviour.
- `pre-commit run` on the changed files, all hooks pass.
- **Not run:** the end-to-end serve tests because they need a larger GPU and I don't have one. In CI, `raw_embeddings_epd` sends a real `file://…safetensors` request. Then, the file is loaded by process_openai_request on that worker, so it covers the success path this PR touches only and not the new error statuses.

### Commands:

```
pre-commit run --files components/src/dynamo/trtllm/multimodal_processor.py \
  components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py
```
```
docker run --rm --gpus all -v "$PWD":/src -w /src -e PYTHONPATH=/src/components/src \
  nvcr.io/nvidia/ai-dynamo/tensorrtllm-runtime:1.4.2 \
  bash -c 'pip install -q pytest==9.0.3 pytest-asyncio==1.3.0 pytest-benchmark==5.2.3 && \
    python -m pytest -p no:cacheprovider -v components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py'
```

New tests:
- `test_embedding_client_errors_return_400`, parametrized over: missing file, outside the allowed dir, too large, missing `mm_embeddings` key, unsupported scheme.
- `test_embedding_outside_allowed_dir_does_not_leak_it`
- `test_embedding_local_path_unconfigured_is_server_error`
- `test_embedding_valid_file_loads`

## Where should the reviewer start?

- `multimodal_processor.py`: `load_tensor_from_path_or_url`, and the embedding `try`/`except` block in `process_openai_request`.
- The status-code choices

## Related Issues

- Closes [#14749](https://github.com/ai-dynamo/dynamo/issues/14749)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved error reporting when loading tensor and embedding files, including clearer validation, size, path, format, and missing-file errors.
  - Added safer handling for local media access and unsupported sources.
  - Prevented sensitive source details from being exposed in error messages.
  - Ensured typed loading errors are correctly surfaced to the frontend while unexpected failures retain generic handling.

### CODE REVIEW

File: components/src/dynamo/trtllm/multimodal_processor.py [hunk 1]
```diff
@@ -458,21 +481,30 @@ async def process_openai_request(
...
                                 if emb is None:
+                                    source = describe_media_source(path)
                                     logging.error(
-                                        "Dictionary embeddings missing 'mm_embeddings' key"
+                                        f"Dictionary embeddings missing 'mm_embeddings' key: {path}"
+                                    )
+                                    raise HttpStatusError(
+                                        400,
+                                        f"Malformed embedding file {path}: missing 'mm_embeddings' key.",
+                                        source,
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/multimodal_processor.py]: Fixed in `b88069a`

File: components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py [hunk 2]
```diff
@@ -542,3 +557,134 @@ async def _encode(_req):
...
+
+def _missing_mm_embeddings_key(tmp_path, processor):
+    """Return a two-key file, loaded as a dict, that lacks the mm_embeddings key."""
+    path = tmp_path / "two_keys.safetensors"
+    save_file({"a": torch.zeros(2), "b": torch.zeros(2)}, str(path))
+    return f"file://{path}"
+
+
+def _unsupported_scheme(tmp_path, processor):
+    """Return a .safetensors URL with a scheme other than http or https."""
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py]: Removed in 5fc1b9b

File: components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py [hunk 3]
```diff
@@ -542,3 +557,134 @@ async def _encode(_req):
...
+    allowed = tmp_path / "allowed"
+    allowed.mkdir()
+    processor.allowed_local_media_path = str(allowed)
+    outside = tmp_path / "outside.safetensors"
+    save_file({"emb": torch.zeros(2)}, str(outside))
+    return f"file://{outside}"
+
+
+def _file_too_large(tmp_path, processor):
+    """Lower the size limit to 1 byte and return a valid file that exceeds it."""
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py]: Removed in 5fc1b9b

File: components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py [hunk 4]
```diff
@@ -542,3 +557,134 @@ async def _encode(_req):
...
+
+# Each setup receives a processor whose allowed_local_media_path is tmp_path,
+# prepares one failure, and returns the URL the client would send.
+def _missing_file(tmp_path, processor):
+    """Return a URL for a file that does not exist inside the allowed directory."""
+    return f"file://{tmp_path / 'missing.safetensors'}"
+
+
+def _outside_allowed_dir(tmp_path, processor):
+    """Restrict the allowed directory to a subfolder and return a file outside it."""
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py]: Removed in 5fc1b9b

File: components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py [hunk 5]
```diff
@@ -542,3 +557,134 @@ async def _encode(_req):
...
+        "multi_modal_data": {"image_url": [{"Url": url}]},
+        "token_ids": [1],
+        "extra_args": {"formatted_prompt": "describe"},
+    }
+
+
+# Each setup receives a processor whose allowed_local_media_path is tmp_path,
+# prepares one failure, and returns the URL the client would send.
+def _missing_file(tmp_path, processor):
+    """Return a URL for a file that does not exist inside the allowed directory."""
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py]: Removed in 5fc1b9b

File: components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py [hunk 6]
```diff
@@ -542,3 +557,134 @@ async def _encode(_req):
...
+        await processor.process_openai_request(
+            _embedding_request(f"file://{path}"),
+            embeddings=None,
+            ep_disaggregated_params=None,
+        )
+
+    assert exc_info.value.status == 500
+
+
+@pytest.mark.asyncio
```
[the remark this answers is not in the document]
Comment (Bossixd) [reply components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py]: Removed in 5fc1b9b
